pkg/destroy/gcp/network.go

```diff
@@ -0,0 +1,647 @@
+	}
+	if len(candidates) == 0 {
+		return nil, nil
+	}
+
+	igName := ""
+	for _, ig := range candidates {
+		instances, err := o.listInstanceGroupInstances(ig)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get internal LB instance group instances for %s", ig.name)
+		}
+		if o.areAllClusterInstances(instances) {
+			igName = ig.name
+			break
+		}
+	}
+	igs := []nameAndZone{}
+	if len(igName) > 0 {
+		for _, ig := range candidates {
+			if ig.name == igName {
+				igs = append(igs, ig)
+			}
+		}
+	}
+
+	return igs, nil
+}
+
+func (o *ClusterUninstaller) areAllClusterInstances(instances []nameAndZone) bool {
+	for _, instance := range instances {
+		if !o.isClusterResource(instance.name) {
+			return false
+		}
+	}
+	return true
+}
+
+func (o *ClusterUninstaller) getInstanceGroupURL(ig nameAndZone) string {
+	return fmt.Sprintf("%s%s/zones/%s/instanceGroups/%s", o.computeSvc.BasePath, o.ProjectID, ig.zone, ig.name)
+}
+
+func (o *ClusterUninstaller) listBackendServicesForInstanceGroups(igs []nameAndZone) ([]string, error) {
+	urls := sets.NewString()
+	for _, ig := range igs {
+		urls.Insert(o.getInstanceGroupURL(ig))
+	}
+	return o.listBackendServices("items(name,backends)", "", func(item *compute.BackendService) bool {
+		if len(item.Backends) == 0 {
+			return false
+		}
+		for _, backend := range item.Backends {
+			if !urls.Has(backend.Group) {
+				return false
+			}
+		}
+		return true
+	})
+}
+
+// deleteInternalLoadBalancer follows a similar cleanup procedure as:
+// https://github.com/openshift/kubernetes/blob/1e5983903742f64bca36a464582178c940353e9a/pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go#L222
+// TODO: add cleanup for shared mode resources (determine if it's supported in 4.2)
+func (o *ClusterUninstaller) deleteInternalLoadBalancer(clusterID string, loadBalancerName string) error {
+	if err := o.deleteAddress(loadBalancerName); err != nil {
+		return err
+	}
+	if err := o.deleteForwardingRule(loadBalancerName); err != nil {
+		return err
+	}
+	// TODO: Figure out a way to preserve the backend service name after this is
+	// gone. Otherwise we can't find this internal load balancer again. However,
+	// the backend service must be deleted first before health checks.
+	if err := o.deleteBackendService(loadBalancerName); err != nil {
+		return err
+	}
+	if err := o.deleteHealthCheck(loadBalancerName); err != nil {
+		return err
+	}
+	if err := o.deleteHealthCheck(loadBalancerName + "-hc"); err != nil {
+		return err
+	}
+	if err := o.deleteFirewall(loadBalancerName + "-hc"); err != nil {
+		return err
+	}
+	return nil
+}
+
+// deleteExternalLoadBalancer follows a similar cleanup procedure as:
+// https://github.com/openshift/kubernetes/blob/1e5983903742f64bca36a464582178c940353e9a/pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go#L289
+// TODO: cleanup nodes health check (using clusterid)
+func (o *ClusterUninstaller) deleteExternalLoadBalancer(loadBalancerName string) error {
```

Comment:
would it be useful to convert these resources to something that our cleanup code for these resource type picks up rather than doing a synchronized delete here...?

Reply:
See my comment above ... not sure how. If you have ideas, please let me know.

Reply:
hmm, that's sad.. looks like this will do for now.